virtcontainers/qemu.go

```diff
@@ -573,6 +575,23 @@ func (q *qemu) startSandbox(timeout int) error {
 		}
 	}()
 
+	if q.config.VirtioFS {
+		sockPath, err := utils.BuildSocketPath(runStoragePath, q.id, "vhost-fs.sock")
+		if err != nil {
+			return err
+		}
+
+		// The daemon will terminate when the vhost-user socket
+		// connection with QEMU closes.  Therefore we do not keep track
+		// of this child process.
+		sourcePath := filepath.Join(kataHostSharedDir, q.id)
+
+		cmd := exec.Command(q.config.VirtioFSDaemon, "-o", "virtio_socket=" + sockPath, "-o", "source=" + sourcePath, "/")
```

Comment:
This part (spawning the `virtiofsd` binary) is not QEMU specific. It can be shared across hypervisors as long as they support `virtio-fs`. That's why it'd be better to have it implemented from hypervisor.go directly so that each hypervisor implementation could reuse the common code. Or, to invoke this from the generic `sandbox.go` code.

Reply:
I looked into this but found sandbox.go and hypervisor.go don't offer convenient places to put this code.  The issue is that qemu.go creats the vm directory so sandbox.go cannot launch the daemon before calling hypervisor.startSandbox().  And putting this function into hypervisor.go is awkward because it relies on paths that come from qemu.go (or another caller who has access to objects).

Let's leave it in qemu.go until there is a need to move it out for Firecracker.